### ISSUE TITLE
Boogie with CVC5

### ISSUE THREAD

**Issue by Anjiang-Wei** [boogie-org/boogie#518]
Does Boogie support the latest version of CVC5 (https://cvc5.github.io/downloads.html)?
I downloaded CVC5 (more specifically, cvc5 version 0.0.7 [git tag 0.0.7 branch HEAD] compiled with GCC version 9.3.0), then ran the tests in the following way, but it did not work:
- `lit -v . --param boogie_params='/proverOpt:SOLVER=CVC5'`  --> no passing test, it seems to be hanging all the time
- `lit -v . --param boogie_params='/proverOpt:PROVER_PATH=XXXXXXX/cvc5/cvc5 # my path to CVC5 binary`  --> mt2 error like below:
```
+Prover error: (error "Error in option parsing: Argument 'mt2' for uint64_t option seed did not parse as uint64_t")

-Boogie program verifier finished with 1 verified, 0 errors
+Prover error:
+
+Prover error: Please use --help to get help on command-line options.
+
+Advisory: MouCreateClassObject SKIPPED due to I/O exception: Broken pipe
+MouCreateClassObject.bpl(3759,12): Verification encountered solver exception (MouCreateClassObject)
+
+Boogie program verifier finished with 0 verified, 0 errors, 1 solver exceptions

error: command failed with exit status: 1
```

This is weird because `mt2` does not seem to appear in the source code of boogie.

I also tried running one single test (sequences/intseq.bpl) for CVC5:
- `lit -v sequences/intseq.bpl --param boogie_params='/proverOpt:SOLVER=CVC5'` It hangs, like above.
- `lit -v sequences/intseq.bpl --param boogie_params='/proverOpt:PROVER_PATH=XXXXXXX/cvc5/cvc5` It raises mt2 error like above.

**Comment by Anjiang-Wei** [comment]
If it is hard to fix, is there any hint for me to debug this issue? Also, it would be greatly appreciated if someone could also reproduce the issue:)

**Comment by Anjiang-Wei** [comment]
I kind of know what's happening, but still unresolved.
- `boogie /proverOpt:SOLVER=cvc5 sequences/intseq.bpl /proverLog:temp.smt2` is the right command to generate smt2 file for debugging. Directly running `cvc5 temp.smt2` raises the following error:
```
(error "Cannot push when not solving incrementally (use --incremental)")
```

After run `cvc5 temp.smt2 --incremental`, it gives
```
unsat
unsupported
```
- specifying `PROVER_PATH` does not even generate smt2 file (i.e., `temp.smt2`), so I still cannot debug the `mt2` case.

**Comment by Anjiang-Wei** [comment]
The specific commit causing the problem has been found:
https://github.com/boogie-org/boogie/commit/da30caa77454a296878056ae5ed7dd4414a9f2ea
It seems that from that specific commit makes boogie always send `(get-info :rlimit)` to the solver, which is only support by Z3 (but not by CVC5).
@[atomb](https://github.com/boogie-org/boogie/commits?author=atomb) Is there an easy way to fix so that it does not send `(get-info :rlimit)` when using CVC5?

**Comment by atomb** [comment]
Yes, I think it would be relatively easy to make it only emit that command when using Z3. I'll look into that.

Just out of curiosity, if you log the `smt2` file with `/proverLog`, and then remove the `(get-info :rlimit)` commands from the resulting file, does everything else work? I've had very little luck getting Boogie to work with CVC5, myself, even before adding the rlimit support, but I think it's somewhat dependent on the sort of Boogie program you're analyzing. In particular, I've mostly tried it on Boogie programs generated from Dafny.

**Comment by Anjiang-Wei** [comment]
自动回复：您发的邮件已经收到，谢谢。

**Comment by Anjiang-Wei** [comment]
> if you log the smt2 file with /proverLog, and then remove the (get-info :rlimit) commands from the resulting file, does everything else work?

Yes, it works. (No `unsupported` reported any more).
I think that Boogie should be designed to be solver-agnostic. I have tried using a older SHA (`tag: v2.11.2`) of Boogie and Boogie is working quite well with CVC5 at that SHA (i.e., boogie tests pass, and boogie programs generated Dafny also pass).

**Comment by atomb** [comment]
Draft PR #521 fixes this issue along with some related ones, and introduces CVC5 into CI. I'm going to wait until #511 is merged before getting reviews on #521, however, to make the diffs easier to digest.
